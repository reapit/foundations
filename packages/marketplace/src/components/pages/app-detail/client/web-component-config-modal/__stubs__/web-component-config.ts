import { ReduxState } from '@/types/core'

export const webComponentStub = {
  client: {
    webComponent: { loading: false, updating: false, data: null, isShowModal: true, negotiators: {} },
  },
} as ReduxState
