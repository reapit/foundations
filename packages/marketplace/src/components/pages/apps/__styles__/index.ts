import { css } from 'linaria'

export const appList = css`
  width: auto;
  -ms-overflow-style: none;
  scrollbar-width: none;
  margin: -1rem;
  padding: 1rem;
  &::-webkit-scrollbar {
    display: none;
  }
`
