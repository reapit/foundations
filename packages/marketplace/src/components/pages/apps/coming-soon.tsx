import * as React from 'react'
import { GridThreeColItem, Grid } from '@reapit/elements'
import placeHolderImage from '@/assets/images/default-feature-image.jpg'
import comingSoonImage1 from '@/assets/images/coming-soon/1Area.jpg'
import comingSoonImage2 from '@/assets/images/coming-soon/2Starberry.jpg'
import comingSoonImage3 from '@/assets/images/coming-soon/3IAMProperty.jpg'
import comingSoonImage4 from '@/assets/images/coming-soon/4SMS.jpg'
import comingSoonImage5 from '@/assets/images/coming-soon/5Vyomm.jpg'
import comingSoonImage6 from '@/assets/images/coming-soon/6Spectre.jpg'
import comingSoonImage7 from '@/assets/images/coming-soon/7Twentyci.jpg'
import comingSoonImage8 from '@/assets/images/coming-soon/8Zero.jpg'
import comingSoonImage9 from '@/assets/images/coming-soon/9Yomdel.jpg'

export type ComingSoonAppsProps = {
  setComingSoonAppSectionHeight?: React.Dispatch<React.SetStateAction<number>>
}

export const onImageError = (event: React.SyntheticEvent<HTMLImageElement>) =>
  (event.currentTarget.src = placeHolderImage)

const comingSoonImagesList = [
  comingSoonImage1,
  comingSoonImage2,
  comingSoonImage3,
  comingSoonImage4,
  comingSoonImage5,
  comingSoonImage6,
  comingSoonImage7,
  comingSoonImage8,
  comingSoonImage9,
]

const ComingSoonApps: React.FC<ComingSoonAppsProps> = ({ setComingSoonAppSectionHeight }) => {
  const comingSoonAppSectionRef = React.useRef<HTMLDivElement>(null)

  React.useLayoutEffect(() => {
    const offsetHeight = comingSoonAppSectionRef.current?.offsetHeight
    const clientHeight = comingSoonAppSectionRef.current?.clientHeight
    const scrolHeight = comingSoonAppSectionRef.current?.scrollHeight

    const boundingClientRect = comingSoonAppSectionRef.current?.getBoundingClientRect()
    console.log('boundingClientRect', boundingClientRect)

    console.log(
      `useLayoutEffect - ComingSoonApps component -> offsetHeight - ${offsetHeight} - clientHeight - ${clientHeight} - scrolHeight - ${scrolHeight}`,
    )
    if (offsetHeight && setComingSoonAppSectionHeight) {
      setComingSoonAppSectionHeight(offsetHeight)
    }
  })

  return (
    <div ref={comingSoonAppSectionRef}>
      <Grid isMultiLine>
        {comingSoonImagesList.map(imgSrc => (
          <GridThreeColItem key={imgSrc}>
            <div className="card">
              <div className="card-image">
                <img className="image" src={imgSrc} onError={onImageError} />
              </div>
            </div>
          </GridThreeColItem>
        ))}
      </Grid>
    </div>
  )
}

export default ComingSoonApps
