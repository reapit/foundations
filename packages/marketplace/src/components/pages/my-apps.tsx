import * as React from 'react'
import { connect } from 'react-redux'
import { withRouter, RouteComponentProps } from 'react-router'
import { ReduxState, FormState } from '@/types/core'
import { MyAppsState } from '@/reducers/my-apps'
import { Loader, Info } from '@reapit/elements'
import ErrorBoundary from '@/components/hocs/error-boundary'
import routes from '@/constants/routes'
import AppList from '@/components/ui/app-list'
import { appDetailRequestData } from '@/actions/app-detail'
import { AppDetailState } from '@/reducers/app-detail'
import AppDetailModal from '../ui/app-detail-modal'
import { myAppsRequestData } from '@/actions/my-apps'
import { selectClientId } from '@/selector/client'
import { AppSummaryModel } from '@reapit/foundations-ts-definitions'
import { handleLaunchApp } from '../../utils/launch-app'
import { appInstallationsSetFormState } from '@/actions/app-installations'
import { setAppDetailModalStateManage } from '@/actions/app-detail-modal'
import { selectIsAdmin } from '@/selector/auth'

export interface MyAppsMappedActions {
  setStateViewManage: () => void
  fetchAppDetail: (id: string, clientId: string) => void
  fetchMyApp: (page: number) => void
  installationsSetFormState: (formState: FormState) => void
}

export interface MyAppsMappedProps {
  myAppsState: MyAppsState
  appDetail: AppDetailState
  installationsFormState: FormState
  clientId: string
  isAdmin: boolean
}

export type MyAppsProps = MyAppsMappedActions & MyAppsMappedProps & RouteComponentProps<{ page?: any }>

export const handleOnChange = history => (page: number) => history.push(`${routes.MY_APPS}/${page}`)

export const handleUseEffect = ({ isDone, installationsSetFormState, fetchMyApp, pageNumber }) => () => {
  if (isDone) {
    installationsSetFormState('PENDING')
    fetchMyApp(pageNumber)
  }
}

export const MyApps: React.FunctionComponent<MyAppsProps> = ({
  myAppsState,
  clientId,
  match,
  fetchMyApp,
  fetchAppDetail,
  appDetail,
  setStateViewManage,
  installationsFormState,
  installationsSetFormState,
  history,
  isAdmin,
}) => {
  const pageNumber = match.params && !isNaN(match.params.page) ? Number(match.params.page) : 1
  const unfetched = !myAppsState.myAppsData
  const loading = myAppsState.loading
  const list = myAppsState?.myAppsData?.data?.data || []
  const { totalCount, pageSize } = myAppsState?.myAppsData?.data || {}
  const [visible, setVisible] = React.useState(false)
  const isDone = installationsFormState === 'DONE'

  React.useEffect(handleUseEffect({ isDone, installationsSetFormState, fetchMyApp, pageNumber }), [isDone])

  if (unfetched || loading) {
    return <Loader />
  }
  if (!isAdmin) return <Info infoType="404" />

  return (
    <ErrorBoundary>
      <AppList
        list={list}
        title="Manage Apps"
        loading={loading}
        onCardClick={(app: AppSummaryModel) => handleLaunchApp(app)}
        onSettingsClick={(app: AppSummaryModel) => {
          setVisible(true)
          setStateViewManage()
          if (app.id && (!appDetail.appDetailData || appDetail.appDetailData.data.id !== app.id)) {
            fetchAppDetail(app.id, clientId)
          }
        }}
        infoType="INSTALLED_APPS_EMPTY"
        pagination={{
          totalCount,
          pageSize,
          pageNumber,
          onChange: handleOnChange(history),
        }}
      />
      <AppDetailModal visible={visible} afterClose={() => setVisible(false)} />
    </ErrorBoundary>
  )
}

export const mapStateToProps = (state: ReduxState): MyAppsMappedProps => ({
  myAppsState: state.myApps,
  appDetail: state.appDetail,
  clientId: selectClientId(state),
  installationsFormState: state.installations.formState,
  isAdmin: selectIsAdmin(state),
})

export const mapDispatchToProps = (dispatch: any): MyAppsMappedActions => ({
  setStateViewManage: () => dispatch(setAppDetailModalStateManage()),
  fetchAppDetail: (id: string, clientId: string) => dispatch(appDetailRequestData({ id, clientId })),
  fetchMyApp: (page: number) => dispatch(myAppsRequestData(page)),
  installationsSetFormState: (formState: FormState) => dispatch(appInstallationsSetFormState(formState)),
})

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(MyApps))
