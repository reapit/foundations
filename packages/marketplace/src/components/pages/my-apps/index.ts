import MyApps from './my-apps'
export default MyApps
