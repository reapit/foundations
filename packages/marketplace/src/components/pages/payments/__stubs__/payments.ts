export const paymentsDataStub = {
  data: [
    {
      id: 'RPT201231231',
      created: '2019-01-25T15:44:28.0000000Z',
      modified: '2019-01-25T15:44:28.0000000Z',
      type: 'paymentRequest',
      status: 'pending',
      ledger: 'landlord',
      amount: 1000,
      clientAccountName: 'General',
      externalReference: 'myApayaRef',
      contactId: 'OXF18000011',
      propertyId: 'OXF120022',
    },
    {
      id: 'RPT201231232',
      created: '2019-01-25T15:44:28.0000000Z',
      modified: '2019-01-25T15:44:28.0000000Z',
      type: 'paymentRequest',
      status: 'pending',
      ledger: 'landlord',
      amount: 2000,
      clientAccountName: 'General1',
      externalReference: 'myApayaRef',
      contactId: 'OXF18000012',
      propertyId: 'OXF120023',
    },
  ],
  pageNumber: 1,
  pageSize: 2,
  pageCount: 2,
  totalCount: 6,
  isLoading: false,
}
