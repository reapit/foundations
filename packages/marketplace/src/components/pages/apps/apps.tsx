import * as React from 'react'
import { TransitionGroup } from 'react-transition-group'
import { Dispatch } from 'redux'
import qs from 'query-string'
import { History } from 'history'
import { useSelector, useDispatch } from 'react-redux'
import { Loader, Section, H3, Grid } from '@reapit/elements'
import ErrorBoundary from '@/components/hocs/error-boundary'
import { useHistory, useLocation } from 'react-router'
import AppList from '@/components/ui/app-list'
import FeaturedApp from '@/components/ui/featured-app'
// Commenting out as we are disabling for launch because there are too few apps
// import AppSidebar from '@/components/ui/app-sidebar'
import { selectAppsListState, selectFeatureAppsListState } from '@/selector/apps'
import { AppSummaryModel } from '@reapit/foundations-ts-definitions'
import { addQuery, hasFilterParams } from '@/utils/client-url-params'
import Routes from '@/constants/routes'
import InfiniteScroll from 'react-infinite-scroll-component'
import { fetchApps } from '@/actions/apps'
import { getNumberOfItems } from '@/utils/browse-app'
import ComingSoonApps from './coming-soon'
import styles from '@/styles/pages/apps.scss?mod'

export const handleAfterClose = ({ setVisible }) => () => setVisible(false)
export const handleOnChange = history => (page: number) => {
  history.push(addQuery({ page }))
}

export const handleOnCardClick = (history: History) => (app: AppSummaryModel) => {
  history.push(`${Routes.APPS}/${app.id}`)
}

export const handleLoadMore = ({
  dispatch,
  preview,
  loading,
  numOfItemsPerPage,
  pageNumber,
}: {
  dispatch: Dispatch
  preview: boolean
  loading: boolean
  numOfItemsPerPage: number
  pageNumber: number
}) => () => {
  !loading &&
    dispatch(fetchApps({ pageNumber: pageNumber + 1, preview, isInfinite: true, pageSize: numOfItemsPerPage }))
}

export const Apps: React.FunctionComponent = () => {
  const [comingSoonAppSectionHeight, setComingSoonAppSectionHeight] = React.useState(0)
  const history = useHistory()
  const location = useLocation()
  const dispatch = useDispatch()

  const appsListState = useSelector(selectAppsListState)
  const hasParams = hasFilterParams(location.search)
  const apps = appsListState?.data || []
  const { totalCount = 0, pageNumber = 1 } = appsListState || {}

  const featureAppsListState = useSelector(selectFeatureAppsListState)
  const featuredApps = featureAppsListState?.data || []
  const { preview: previewString } = qs.parse(location.search)
  const preview = !!previewString

  const numOfItemsPerPage = getNumberOfItems()
  const totalPage = totalCount / numOfItemsPerPage

  const loading = appsListState.isLoading || featureAppsListState.isLoading
  /**
   * When apps is empty or when loading app set hasMore = false to prevent trigger load more
   * Otherwise set hasMore = true when pageNumber (current page ) less then totalPage
   *
   */
  const hasMore = apps.length == 0 || loading ? false : pageNumber < totalPage

  return (
    <ErrorBoundary>
      <Section
        className={styles['app-list']}
        isFlex
        isFlexColumn
        hasPadding={false}
        hasMargin={false}
        hasBackground={false}
      >
        {/* <AppSidebar /> */}
        <H3 isHeadingSection>Browse Apps</H3>
        {!hasParams && featuredApps.length > 0 && (
          <div className="pb-4 bb mb-4">
            <Grid isMultiLine>
              {featuredApps.map(app => (
                <FeaturedApp key={app.id} app={app} />
              ))}
            </Grid>
          </div>
        )}
        <InfiniteScroll
          dataLength={apps.length}
          next={handleLoadMore({ dispatch, preview, loading, numOfItemsPerPage, pageNumber })}
          hasMore={hasMore}
          loader={<Loader key="infiniteScrollLoader" />}
          scrollThreshold={`${comingSoonAppSectionHeight}px`}
          // We disable the scrolling in the app list  container and allow the app root container to scroll
          // so the scrollableTarget must be set as app-root-container
          scrollableTarget="app-root-container"
        >
          <TransitionGroup>
            <AppList
              list={apps}
              loading={loading}
              onCardClick={handleOnCardClick(history)}
              infoType={pageNumber > 1 || hasParams ? '' : 'CLIENT_APPS_EMPTY'}
              animated
            />
          </TransitionGroup>
        </InfiniteScroll>

        <div className="bb mb-4" />
        <ComingSoonApps setComingSoonAppSectionHeight={setComingSoonAppSectionHeight} />
      </Section>
    </ErrorBoundary>
  )
}

export default Apps
