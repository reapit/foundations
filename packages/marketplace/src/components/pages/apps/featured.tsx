import React from 'react'
import Slider from 'react-slick'
import { AppSummaryModel } from '@reapit/foundations-ts-definitions'
import FeaturedApp from '../../ui/featured-app'
import { SlickWrapper } from './__styles__/slick'
import { Section } from '@reapit/elements-legacy'

export interface FeaturedAppsProps {
  apps: AppSummaryModel[]
}

export const FeaturedApps: React.FC<FeaturedAppsProps> = ({ apps }) => {
  while (apps.length < 5) {
    apps.push({})
  }

  const settings = {
    dots: true,
    infinite: true,
    speed: 500,
    className: 'center',
    centerMode: true,
    centerPadding: '60px',
    slidesToShow: 4,
    initialSlide: 0,
    responsive: [
      {
        breakpoint: 1600,
        settings: {
          slidesToShow: 3,
          slidesToScroll: 3,
        },
      },
      {
        breakpoint: 1215,
        settings: {
          slidesToShow: 2,
          slidesToScroll: 2,
        },
      },
      {
        breakpoint: 850,
        settings: {
          slidesToShow: 1,
          slidesToScroll: 1,
        },
      },
    ],
  }
  return (
    <Section>
      <SlickWrapper>
        <Slider {...settings}>
          {apps.map((app, index) => {
            return <FeaturedApp key={app.id || index} app={app} />
          })}
        </Slider>
      </SlickWrapper>
    </Section>
  )
}
