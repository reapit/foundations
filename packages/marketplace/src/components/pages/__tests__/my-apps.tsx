import * as React from 'react'
import { shallow } from 'enzyme'

import { appsDataStub } from '@/sagas/__stubs__/apps'
import { MyAppsItem } from '@/reducers/my-apps'
import routes from '@/constants/routes'
import { MyApps, MyAppsProps, handleOnChange, handleUseEffect, mapStateToProps, mapDispatchToProps } from '../my-apps'
import { ReduxState } from '@/types/core'

const mockProps = (loading: boolean, appData: MyAppsItem | null, isAdmin: boolean): MyAppsProps => ({
  myAppsState: {
    loading: loading,
    myAppsData: appData,
  },
  // @ts-ignore: just pick the needed props for the test
  match: {
    params: {
      page: '2',
    },
  },
  isAdmin,
})

describe('MyApps', () => {
  it('should match a snapshot when LOADING false', () => {
    expect(shallow(<MyApps {...mockProps(false, appsDataStub, true)} />)).toMatchSnapshot()
  })

  it('should match a snapshot when LOADING true', () => {
    expect(shallow(<MyApps {...mockProps(true, null, true)} />)).toMatchSnapshot()
  })

  it('should match a snapshot when isAdmin false', () => {
    expect(shallow(<MyApps {...mockProps(false, null, false)} />)).toMatchSnapshot()
  })

  it('handleOnChange', () => {
    const mockHistory = {
      push: jest.fn(),
    }
    const fn = handleOnChange(mockHistory)
    fn(1)
    expect(mockHistory.push).toBeCalledWith(`${routes.MY_APPS}/${1}`)
  })

  it('handleUseEffect', () => {
    const mockProps = {
      isDone: true,
      installationsSetFormState: jest.fn(),
      fetchMyApp: jest.fn(),
      pageNumber: 1,
    }
    const fn = handleUseEffect(mockProps)
    fn()
    expect(mockProps.installationsSetFormState).toBeCalled()
    expect(mockProps.fetchMyApp).toBeCalledWith(mockProps.pageNumber)
  })

  it('mapStateToProps', () => {
    const mockState = {
      myApps: {},
      appDetail: {},
      auth: {
        loginSession: {
          loginIdentity: {
            clientId: '',
            isAdmin: true,
          },
        },
      },
      installations: {
        formState: 'PENDING',
      },
    } as ReduxState
    const result = mapStateToProps(mockState)
    const output = {
      myAppsState: {},
      appDetail: {},
      clientId: '',
      installationsFormState: 'PENDING',
      isAdmin: true,
    }
    expect(result).toEqual(output)
  })

  describe('mapDispatchToProps', () => {
    it('fetchAppDetail', () => {
      const mockDispatch = jest.fn()
      const { fetchAppDetail } = mapDispatchToProps(mockDispatch)
      fetchAppDetail('123', '')
      expect(mockDispatch).toBeCalled()
    })

    it('fetchMyApp', () => {
      const mockDispatch = jest.fn()
      const { fetchMyApp } = mapDispatchToProps(mockDispatch)
      fetchMyApp(1)
      expect(mockDispatch).toBeCalled()
    })

    it('installationsSetFormState', () => {
      const mockDispatch = jest.fn()
      const { installationsSetFormState } = mapDispatchToProps(mockDispatch)
      installationsSetFormState('PENDING')
      expect(mockDispatch).toBeCalled()
    })
  })
})
