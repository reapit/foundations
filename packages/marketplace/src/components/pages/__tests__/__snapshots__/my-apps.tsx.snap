// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyApps should match a snapshot when LOADING false 1`] = `
<Connect(ErrorBoundary)>
  <Component
    infoType="INSTALLED_APPS_EMPTY"
    list={
      Array [
        Object {
          "created": "2020-02-02T10:45:57",
          "developer": "Pete's Proptech World Ltd",
          "developerId": "28c9ea52-7f73-4814-9e00-4e3714b8adeb",
          "homePage": "http://google.com/abc",
          "iconUri": "https://reapit-app-store-app-media.s3.eu-west-2.amazonaws.com/d10e790c-2bf2-40ae-9c43-82c1534bde31.png",
          "id": "09043eb8-9e5e-4650-b7f1-f0cb62699027",
          "links": Array [
            Object {
              "action": "GET",
              "href": "http://platformdemo.reapit.net/marketplace/apps/09043eb8-9e5e-4650-b7f1-f0cb62699027",
              "rel": "self",
            },
            Object {
              "action": "GET",
              "href": "http://platformdemo.reapit.net/marketplace/developers/28c9ea52-7f73-4814-9e00-4e3714b8adeb",
              "rel": "developer",
            },
          ],
          "name": "test",
          "summary": "nXXT2zaK807ysWgy8F0WEhIYRP3TgosAtfuiLtQCImoSx0kynxbIF0nkGHU36Oz13kM3DG0Bcsicr8L6eWFKLBg4axlmiOEWcvwHAbBP9LRvoFkCl58k1wjhOExnpaZItEyOT1AXVKv8PE44aMGtVz",
        },
        Object {
          "created": "2020-02-02T10:45:57",
          "developer": "Pete's Proptech World Ltd",
          "developerId": "28c9ea52-7f73-4814-9e00-4e3714b8adeb",
          "homePage": "http://www.contoso.com/path",
          "iconUri": "https://reapit-app-store-app-media.s3.eu-west-2.amazonaws.com/1d38c5ec-e3c0-4138-9fe0-e77bce4034d1.octet-stream",
          "id": "261da083-cee2-4f5c-a18f-8f9375f1f5af",
          "links": Array [
            Object {
              "action": "GET",
              "href": "http://platformdemo.reapit.net/marketplace/apps/261da083-cee2-4f5c-a18f-8f9375f1f5af",
              "rel": "self",
            },
            Object {
              "action": "GET",
              "href": "http://platformdemo.reapit.net/marketplace/developers/28c9ea52-7f73-4814-9e00-4e3714b8adeb",
              "rel": "developer",
            },
          ],
          "name": "asd",
          "summary": "asdQiiAJTmXGxPin7pwUiCsepZWXz1EJS71eGlvgPKH4hpE6J8DRDpzP2kDdOwpQPr4aHCCwWwxBJwzARLa7wMpJh5J61GhmQjLfKZkcDd47L9WEfQYVYAj0DTPJP0BuUMAAg2",
        },
      ]
    }
    loading={false}
    onCardClick={[Function]}
    onSettingsClick={[Function]}
    pagination={
      Object {
        "onChange": [Function],
        "pageNumber": 2,
        "pageSize": 2,
        "totalCount": 6,
      }
    }
    title="Manage Apps"
  />
  <Connect(Component)
    afterClose={[Function]}
    visible={false}
  />
</Connect(ErrorBoundary)>
`;

exports[`MyApps should match a snapshot when LOADING true 1`] = `<Component />`;

exports[`MyApps should match a snapshot when isAdmin false 1`] = `<Component />`;
